File: CacheCSharp/MemoryProvider.cs
using System;
using System.Runtime.Caching;

namespace CacheCSharp
{
    public class MemoryProvider : CacheProviderBase<System.Runtime.Caching.MemoryCache>
    {
        protected override MemoryCache InitCache()
        {
            return MemoryCache.Default;
        }

        public override T Get<T>(string key)
        {
            try
            {
                if (Cache[KeyPrefix + key] == null)
                {
                    return default(T);
                }

                return (T)Cache[KeyPrefix + key];
            }
            catch
            {
                return default(T);
            }
        }

        public override void Set<T>(string key, T value)
        {
            var policy = new CacheItemPolicy();
            Cache.Set(key, value, policy);
        }

        public override void SetSliding<T>(string key, T value)
        {
            SetSliding<T>(KeyPrefix + key, value, CacheDuration);
        }

        public override void Set<T>(string key, T value, int duration)
        {
            var policy = new CacheItemPolicy {AbsoluteExpiration = DateTime.Now.AddMinutes(duration)};
            Cache.Set(key, value, policy);
        }

        public override void SetSliding<T>(string key, T value, int duration)
        {
            var policy = new CacheItemPolicy { SlidingExpiration = new TimeSpan(0, duration, 0) };
            Cache.Set(key, value, policy);
        }

        public override void Set<T>(string key, T value, DateTimeOffset expiration)
        {
            var policy = new CacheItemPolicy { AbsoluteExpiration = expiration.DateTime };
            Cache.Set(key, value, policy);
        }

        public override bool Exists(string key)
        {
            return Cache[key] != null;
        }

        public override void Remove(string key)
        {
            Cache.Remove(KeyPrefix + key);
        }
    }
}

File: CacheVB/MemoryProvider.vb
﻿Imports System.Runtime.Caching

Public Class MemoryProvider
    Inherits CacheProviderBase(Of System.Runtime.Caching.MemoryCache)
    Protected Overrides Function InitCache() As MemoryCache
        Return MemoryCache.[Default]
    End Function

    Public Overrides Function [Get](Of T)(key As String) As T
        Try
            If Not Exists(key) Then
                Return Nothing
            End If

            Return DirectCast(Cache(KeyPrefix & key), T)
        Catch
            Return Nothing
        End Try
    End Function

    Public Overrides Sub [Set](Of T)(key As String, value As T, duration As Integer)
        Dim policy = New CacheItemPolicy()
        policy.AbsoluteExpiration = DateTime.Now.AddMinutes(duration)
        Cache.[Set](KeyPrefix & key, value, policy)
    End Sub

    Public Overrides Sub SetSliding(Of T)(key As String, value As T, duration As Integer)
        Dim policy = New CacheItemPolicy()
        policy.SlidingExpiration = New TimeSpan(0, duration, 0)
        Cache.[Set](KeyPrefix & key, value, policy)
    End Sub

    Public Overrides Sub [Set](Of T)(key As String, value As T, expiration As DateTimeOffset)
        Dim policy = New CacheItemPolicy()
        policy.AbsoluteExpiration = expiration.DateTime
        Cache.[Set](KeyPrefix & key, value, policy)
    End Sub

    Public Overrides Function Exists(key As String) As Boolean
        Return Cache(KeyPrefix & key) IsNot Nothing
    End Function

    Public Overrides Sub Remove(key As String)
        Cache.Remove(KeyPrefix & key)
    End Sub
End Class